### ISSUE TITLE
Improving the API

### ISSUE THREAD

**Issue by franleplant** [erikras/redux-form#478]
Hi!
    Before all, I'd like to thank you for this beautiful library, the best validation library out there, IMO.

I have been using it for a while and a couple of things made me think:

- Why even use redux?

Is there any use for redux here? Why not just use `this.state` to save form state?
It'll make the library need less bootstraping to work.

This should be all you need:

```javascript
@reduxForm({
  form: 'synchronousValidation',
  fields,
  validate
})
class MyClass extends React.Component { ... }
```

- Field list

It will be really nice not needing to specify the `fields` array in order to bootstrap the lib.

I don't know if this is possible but the ideal API should be like this:

```javascript
@reduxForm
class MyClass extends React.Component { ... }
```

The lib would need to be smart enough to detect the inputs that need validation (possibly marked with an attribute or a function call) and do it's job.

> EDIT: I think we can use the `onChange` callback to get the `event.target.name` of the input and filling the `this.props.fields[name]` correctly. So, avoiding the fields array could be technically feasible. The only constraint is that input has a `name` attribute

> EDIT2: [proxies](https://babeljs.io/docs/learn-es2015/#proxies) should be used to achieve this interface. 

- Validators, validation strategies

The validate function is super flexible but most of the time I make the same checks. Granted, you can abstract that into your own code base but that will be duplicated code through different apps using this lib.

I think that we need to implement validators that should work similar to, and have similar API to native ones:

```javascript
<input
  reduxFormValidation={['email']}
/>

<input
  reduxFormValidation={['required email minlength:10']}
/>
```

That syntax could be extended to support custom error messages:

```javascript
<input
  reduxFormValidation={[
  type: 'email',
  customValidator: function validateEmail(value) {}
  message: 'Please use a valid email address dude!'
]}
/>
```

Also it could be in the form of separate attributes instead of an array, it could even hijack default HTML5 validation attributes (that could be awesome)

- Display error messages

Displaying errors if any is verbose:

```javascript
<input type="text" placeholder="Email" {...email}/>
{email.touched && email.error && <div>{email.error}</div>}
```

What about something like this?

```javascript
<input type="text" placeholder="Email" {...email}/>
{email.hasError(<span>{email.error}</span>)} 
```

Does any of these changes make sense for you? Would anybody be interested in these features?
Thanks a lot for your time!
Fran

**Comment by erikras** [comment]
> Why even use redux?

It's certainly true that the state could be kept in the outer component. If there isn't one already, I think the ecosystem has room for a library like that. I used to have a [Benefits](https://github.com/erikras/redux-form/tree/v2.4.5#benefits) section in the `< v3.x` docs that addressed the "Why redux?" question a little.

> Field list

I agree that specifying the array of fields is cumbersome. I would definitely be interested in ideas for how to crawl the element tree to find them.

> Validators, validation strategies

This has been a bit of a contentious point. Some of my views can be found on #65 and #98. I _definitely_ want to look into using the HTML5 validation stuff in the future. (#254)

> Display error messages

Seems like extra API surface area for not much gain, plus your implementation enforces an opinion about when errors should be shown (when `touched`) that I don't think the library should enforce. See https://github.com/erikras/redux-form/issues/98#issuecomment-143534271 (3rd bullet).

Thank you very much for your detailed analysis and brainstorming.

**Comment by yleclanche** [comment]
I disagree for the "display error messages" too.

You can write a small Error component and replace your code with :
```
<input type="text" placeholder="Email" {...email}/>
<Error {...email} />
```
Or even wrap the input in a `<Field>` component and display errors in it so you do not have to rewrite the same code again and again.

**Comment by franleplant** [comment]
> Field list

This could be super cool with Proxies, but the support being as it is, it's not feasible at all.

Im exploring something like this:

```javascript
<input {...fields('email')}
```

Is not ideal, but it could be migrated to proxies easily.

**Comment by bmueller-sykes** [comment]
FWIW, for the forms I've created using redux-form, I abstract as much of the field-specific info as I can into an array of objects, like this:

    const structure = {
	"firstName": { type: "text", label: "First Name", placeholder: "Enter first name..."},
        "lastName": {type: "text", label: "Last Name", placeholder: "enter last name..."},
        etc
    }

...and then I do the same with validation (e.g. `const validation = [...]`). Then I created a React Component called `<FormInput>`, which is just a wrapper for a redux-form-enabled element that sets various default values, validates inbound data, etc, and so then invoking it just looks like this:

<FormInput structure={structure.firstName} data={firstName} />

...where "data" is the data provided by redux-form itself.  Thus far, I've liked this approach since it reduces boilerplate dramatically, and increases readability of JSX/HTML/Form composition.  Not sure if this is a useful strategy for the world at large, but it's worked for me thus far.
